caching-service/src/main/java/org/zowe/apiml/caching/service/infinispan/storage/InfinispanStorage.java

```diff
@@ -10,71 +10,66 @@
 package org.zowe.apiml.caching.service.infinispan.storage;
 
 import lombok.extern.slf4j.Slf4j;
-import org.infinispan.Cache;
 import org.zowe.apiml.caching.model.KeyValue;
 import org.zowe.apiml.caching.service.Messages;
 import org.zowe.apiml.caching.service.Storage;
 import org.zowe.apiml.caching.service.StorageException;
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
 
 @Slf4j
 public class InfinispanStorage implements Storage {
 
 
-    private final Cache<String, Map<String, KeyValue>> cache;
+    private final ConcurrentMap<String, KeyValue> cache;
 
-    public InfinispanStorage(Cache<String, Map<String, KeyValue>> cache) {
+    public InfinispanStorage(ConcurrentMap<String, KeyValue> cache) {
         this.cache = cache;
     }
 
     @Override
     public KeyValue create(String serviceId, KeyValue toCreate) {
+        toCreate.setServiceId(serviceId);
         log.info("Writing record: {}|{}|{}", serviceId, toCreate.getKey(), toCreate.getValue());
 
-        Map<String, KeyValue> serviceCache = cache.computeIfAbsent(serviceId, k -> new HashMap<>());
+        KeyValue serviceCache = cache.putIfAbsent(serviceId + toCreate.getKey(), toCreate);
```

Comment:
Nested map structure was used in the original implementation of Infinispan storage and it turned out to be slow and error-prone in HA setup. Cache object can be mutated by another instance of Caching service when it is not directly used, e.g. ```
Map<String, KeyValue> serviceCache = cache.computeIfAbsent(serviceId, k -> new HashMap<>());
KeyValue entry = serviceCache.put(toCreate.getKey(), toCreate);
        cache.put(serviceId, serviceCache);
``` 
after conputeIfAbsent is returned, the cache lock is released and can be acquired by another instance. When cache.put() is called, the object could not be the same. It could be also an issue for services with too many entries. Each call for serviceId will return a large object that will not be necessarily useful.
There is a trade-off for read/delete all methods but I think that those will not be used extensively. If this will not be the case, we can adjust them.